src/github.com/matrix-org/dendrite/mediaapi/writers/fileutils.go

```diff
@@ -0,0 +1,287 @@
+	if err != nil {
+		return "", -1, -1, "", err
+	}
+	defer tmpFile.Close()
+
+	// The file data is hashed and the hash is returned. The hash is useful as a
+	// method of deduplicating files to save storage, as well as a way to conduct
+	// integrity checks on the file data in the repository. The hash gets used as
+	// the MediaID.
+	hasher := sha256.New()
+
+	// bytesResponded is the total number of bytes written to the response to the client request
+	// bytesWritten is the total number of bytes written to disk
+	var bytesResponded, bytesWritten int64 = 0, 0
+	var bytesTemp int64
+	var copyError error
+	// Note: the buffer size is the same as is used in io.Copy()
+	buffer := make([]byte, 32*1024)
+	for {
+		// read from remote request's response body
+		bytesRead, readErr := reqReader.Read(buffer)
+		if bytesRead > 0 {
+			// Note: This code allows proxying files larger than maxFileSizeBytes!
+			// write to client request's response body
+			bytesTemp, copyError = writeToResponse(respWriter, buffer, bytesRead)
+			bytesResponded += bytesTemp
+			if copyError == nil {
+				// Note: if we get here then copyError != errFileIsTooLarge && copyError != errWrite
+				//   as if copyError == errResponse || copyError == errWrite then we would have broken
+				//   out of the loop and there are no other cases
+				bytesTemp, copyError = writeToDiskAndHasher(tmpFileWriter, hasher, bytesWritten, maxFileSizeBytes, buffer, (bytesRead))
+				bytesWritten += bytesTemp
+				// If we do not have a respWriter then we are only writing to the hasher and tmpFileWriter. In that case, if we get an error, we need to break.
+				if respWriter == nil && copyError != nil {
+					break
+				}
+			}
+		}
+		if readErr != nil {
+			if readErr != io.EOF {
+				copyError = errRead
+			}
+			break
+		}
+	}
+
+	if copyError != nil {
+		return "", -1, -1, "", copyError
+	}
+
+	tmpFileWriter.Flush()
+
+	hash := hasher.Sum(nil)
+	return types.Base64Hash(base64.URLEncoding.EncodeToString(hash[:])), types.ContentLength(bytesResponded), types.ContentLength(bytesWritten), tmpDir, nil
+}
+
+// getPathFromMediaMetadata validates and constructs the on-disk path to the media
+// based on its origin and mediaID
+// If a mediaID is too short, which could happen for other homeserver implementations,
+// place it into a short-id subdirectory of the origin directory
+// If the mediaID is long enough, we split it into pieces, creating up to 2 subdirectories
+// for more manageable browsing and use the remainder as the file name. For example, if
+// mediaID is 'qwerty', we create subdirectories 'q', 'w' within 'q' and place the file
+// in 'q/w' calling it 'erty'. If the mediaID is shorter than 3 characters, the last
+// character is the file name and the preceding character, if any, is a subdirectory name.
+func getPathFromMediaMetadata(m *types.MediaMetadata, absBasePath types.Path) (string, error) {
+	var subPath, fileName string
+
+	hashLen := len(m.Base64Hash)
```

Comment:
This function only ever uses `m.Base64Hash`. It would be nicer if you just passed that in then rather than the whole metadata `struct`.